File: chop.rb
#!/usr/bin/env ruby
# encoding: UTF-8

def usage(message)
  name = File.basename($PROGRAM_NAME)

  if message
    puts "#{name}: #{message}"
    puts
  end

  puts <<-eos
#{name} --header <HEADER> --wanted <WANTED> --unwanted <UNWANTED> <FILE1> <FILE2> ... <FILEN>
    Processes all the files <FILE1> to <FILEN> and breaks them into blocks on
    any line containing the <HEADER> text

    If <UNWANTED> is given and the block contains it the block will not be displayed

    If <WANTED> is given and the block also contains it the block will be displayed

    If <WANTED> or <UNWANTED> is not given then we use <HEADER> in place of <WANTED>
eos

  exit(1)
end

def find_opts(list, required = [])
  rest = []
  args = {}

  while list.any?
    x = list.shift
    if x.index('--') == 0
      key = x[2..-1].downcase
      usage("Argument #{x} already supplied") if args.key?(key)

      if key.include?('=')
        key, value = key.split('=', 2)
      else
        value = list.shift
      end

      usage("No value given for #{x}") if value.nil?
      args[key] = value
    else
      rest << x
    end
  end

  required.each do |r|
    usage("Required argument --#{r} is missing") unless args.key?(r)
  end

  return args, rest
end

args, rest = find_opts(ARGV, %w(header))

one_of = false
%w(wanted unwanted).each do |arg|
  one_of = true if args.key?(arg)
end

##
# If wanted or unwanted is not given then we use
# the header in place of wanted
##
args['wanted'] = args['header'] unless one_of

def process(text, wanted, unwanted)
  return if wanted && !text.include?(wanted)
  return if unwanted && text.include?(unwanted)
  return if text == ''

  puts text
  puts
end

text = ''

rest.each do |filename|
  File.open(filename, 'r').each do |line|
    if line.include?(args['header'])
      process(text, args['wanted'], args['unwanted'])
      text = ''
    end
    text << line
  end
  process(text, args['wanted'], args['unwanted'])
end

File: chop.nim
import strutils
import os
import tables

var args = initTable[string, string]()
var files: seq[string] = @[]
var text: seq[string] = @[]

proc get_args() =
  var i = 1
  var line, value: string

  while i <= paramCount():
    line = paramStr(i).string

    if line.startsWith("--"):
      line = toLowerAscii(line[2 .. ^1])
      i += 1
      value = paramStr(i).string
      args[line] = value
    else:
      files.add(line)

    i += 1

proc process(text: seq[string]) =
  var found = true

  if text.len() == 0:
    return

  if hasKey(args, "wanted"):
    found = false
    for line in text:
      if contains(line, args["wanted"]):
        found = true
        break

  if hasKey(args, "unwanted"):
    for line in text:
      if line.contains(args["unwanted"]):
        found = false
        break

  if found:
    for line in text:
      echo line
    echo ""

proc add_line(line: string) =
  if contains(line, args["header"]):
    process(text)
    text = @[]
  text.add(line)

get_args()

if args.len() != 2 and args.len() != 3:
  echo "--header is required with at least one of --wanted and / or --unwanted"
  quit(QuitFailure)

if not hasKey(args, "header"):
  echo "Missing argument --header"
  quit(QuitFailure)

var one_of = false
if hasKey(args, "wanted"):
  one_of = true
elif hasKey(args, "unwanted"):
  one_of = true

if not one_of:
  echo "Missing argument --wanted or --unwanted"
  quit(QuitFailure)

if files.len() == 0:
  for line in stdin.lines:
    add_line(line)
  process(text)
else:
  for file in files:
    for line in lines file:
      add_line(line)
    process(text)
